core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java

```diff
@@ -88,14 +93,15 @@ public void stop() {
    public CacheTopology join(String cacheName, CacheJoinInfo joinInfo, CacheTopologyHandler stm,
          PartitionHandlingManager phm) throws Exception {
       log.debugf("Node %s joining cache %s", transport.getAddress(), cacheName);
-      LocalCacheStatus cacheStatus = new LocalCacheStatus(joinInfo, stm, phm);
+      LocalCacheStatus cacheStatus = new LocalCacheStatus(joinInfo, stm, phm, stateTransferExecutor);
       runningCaches.put(cacheName, cacheStatus);
 
       long timeout = joinInfo.getTimeout();
       long endTime = timeService.expectedEndTime(timeout, TimeUnit.MILLISECONDS);
-      // Synchronize here to delay any rebalance until after we have received the initial cache topology.
-      // This ensures that the cache will have a topology at the end of startup (with awaitInitialTransfer disabled).
-      synchronized (cacheStatus) {
+      // Pretend the join elay any rebalance until after we have received the initial cache topology.
+      // This ensures that the initial topology and the GET_CACHE_LISTENERS request will happen on the user thread
+      cacheStatus.getTopologyUpdatesCompletionService().continueTaskInBackground();
```

Comment:
I didn't finish updating the comment... I want to make sure the first topology update is on this thread (presumably a user thread), because installing the first topology implies retrieving the clustered listeners and blocking.